[Shadergraphs] Support for mip sampling modes on 2d textures, 2d texture arrays and 3d textures

kecho opened the pull request:
### Purpose of this PR
Proposing adding support for mip sampling modes on SampleTexture2DNode, SampleTexture2DNodeArray and SampleTexture3DNode.
* Added a new drop down box on these nodes "Mip Sampling Mode". 
* When a mip sampling mode is chosen, new pin comes to live allowing content creator to connect proper inputs.

![image](https://user-images.githubusercontent.com/789671/145892498-d937990b-ae4b-4c51-b683-d2457248013e.png)

Example of LOD, Gradient and Bias

![image](https://user-images.githubusercontent.com/789671/145892607-a5613bda-3eba-4f32-b5a9-6033566e7e5e.png)


---
### Testing status
* Tested with DLSS, global mip bias off and on on nodes.
* Tested live editing
* Tested loading old shader graphs, ensuring the setting is off by default
* Suggesting to QA to test texture array node

---
### Comments to reviewers
CI Tests will come once we get design consensus

bencloward commented:
Instead of using a check box to select the option to use Bias, I would like to see this feature implemented as a drop-down where the user can select Standard (tex2D), Gradient (tex2DGrad), Bias (tex2DBias), or LOD (tex2DLOD), and maybe also Projection (tex2DProj).  This will give the user the full range of mip level selection options that are available in HLSL.

cdxntchou commented:
Remember to test usage in the vertex stage -- only LOD or gradient mode should be allowed there.

Also what should happen if you switch modes while the node is attached to the vertex stage?

kecho commented:
> Remember to test usage in the vertex stage -- only LOD or gradient mode should be allowed there.
> 
> Also what should happen if you switch modes while the node is attached to the vertex stage?

@cdxntchou so right now a texture node does not have any outputs for vertex stage.
![image](https://user-images.githubusercontent.com/789671/145916084-64d01f67-71aa-406e-9756-9fca86c7ac32.png)

In the UI you can't connect anything to the vertex inputs anyway.
Supporting any vertex operations is out of scope of this PR, and should be treated as a separate feature

bencloward commented:
Kleber, Chris's point is really important.  Adding these features means we will want to deprecate the 2D LOD sampler node so we don't have overlapping functionality.  We want this node that you're working on to replace it.  And in order to do that, we need this node to be able to used in the vertex shader.  We don't want to require our users to have to use a different node to sample textures for the vertex shader than what they're accustomed to using for the pixel shader. I hadn't considered this when I went over it with you yesterday, but we do need to make sure we handle that case.  If doing so makes the project too complicated for you, we can go back to our previous plan of waiting until our team makes the node as part of the GTF migration.

cdxntchou commented:
> so right now a texture node does not have any outputs for vertex stage.

Ah, yeah, that would work too -- depends on how you want it to work.

Either we have one SampleTexture2D node that does everything and handles fragment or vertex limitations properly.
Or we have two nodes, one for fragment stage and one for vertex stage.

Two nodes seems easier to implement, but one node is probably a better UX.

kecho commented:
@bencloward yes I think you guys make good points. Not having vertex support here does more damage than good. I'll figure out a way of supporting it, then everything will just work :)

kecho commented:
@cdxntchou @bencloward Success! so I have vertex working just for LOD and Gradient modes.
* Here is what it looks like:
![image](https://user-images.githubusercontent.com/789671/146114301-706ec2c0-0b4f-44d4-a8fb-082ee5dbb738.png)

* In case that you change the mode while connected to vertex to something else, like standard, the shader graph system throws you a warning in the connector
![image](https://user-images.githubusercontent.com/789671/146115388-43e238e5-c729-48e7-9baf-73fe6d562e7e.png)

You also get a bunch of errors in the Console, and the rendering looks invalid
![image](https://user-images.githubusercontent.com/789671/146115439-7590d435-17ff-4d42-83e8-cb305157c930.png)


* Fixed the SampleTexture3D node, to be consistent on its outputs. Added also the swizzle mechanics. Checked and should be backwards compatible since I kept the same slot IDs.
![image](https://user-images.githubusercontent.com/789671/146115144-4bde5fc3-a9f7-48d1-a674-ad95e7db17c4.png)

bencloward reviewed (approved):
I have tested this and found all to be working as designed.  Really looking forward to having these new features!

cdxntchou reviewed (approved):
Looks good!

kecho commented on com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture2DArrayNode.cs:
```diff
@@ -59,7 +77,8 @@ public sealed override void UpdateNodeAfterDeserialization()
             AddSlot(new SamplerStateMaterialSlot(SamplerInput, kSamplerInputName, kSamplerInputName, SlotType.Input));
-            RemoveSlotsNameNotMatching(new[] { OutputSlotRGBAId, OutputSlotRId, OutputSlotGId, OutputSlotBId, OutputSlotAId, TextureInputId, IndexInputId, UVInput, SamplerInput });
+            UpdateMipSamplingModeInputs();
+            RemoveSlotsNameNotMatching(new[] { OutputSlotRGBAId, OutputSlotRId, OutputSlotGId, OutputSlotBId, OutputSlotAId, TextureInputId, IndexInputId, UVInput, SamplerInput, MipBiasInput });
```
Yes! nice catch

kecho commented on com.unity.shadergraph/Documentation~/Sample-Texture-2D-Node.md:
```diff
@@ -29,7 +33,8 @@ If you experience texture sampling errors while using this node in a graph which
 |  Space                   | Dropdown | Tangent, Object | Selects the space of the normal map. If Type is not Normal, this control is ignored. |
-|  Enable Global Mip Bias  | Toggle   | On, Off         | Enables the automatic global mip bias imposed by the runtime. This bias is set during certain dynamic resolution scaling algorithms to improve detail reconstruction. |
+|  Use Global Mip Bias  | Toggle   | On, Off         | Enables the automatic global mip bias imposed by the runtime. This bias is set during certain dynamic resolution scaling algorithms to improve detail reconstruction. |
+|  Mip Sampling Mode  | Dropdown   | Standard, LOD, Gradient, Bias | Chooses different types of mip sampling modes. Standard calculates the mip automatically. LOD mode specifies a mip explicitely. Bias applies a bias (in addition of the global bias). Gradient utilizes explicit gradients for the UVs. |
```
Yes, this global bias comes for DRS. You can disable the global bias though by an already existant checkbox. Ill re-explain this.

kecho commented on com.unity.shadergraph/Documentation~/Sample-Texture-2D-Array-Node.md:
```diff
@@ -27,7 +31,8 @@ NOTE: This [Node](Node.md) can only be used in the **Fragment** shader stage.
 | Name                     | Type     | Options         | Description |
 |:------------------------ |:---------|:----------------|:-------------------------|
-|  Enable Global Mip Bias | Toggle   | On, Off         | Enables the global mipmap bias that Unity automatically imposes at runtime. Unity sets this bias during certain dynamic resolution scaling algorithms to improve detail reconstruction. |
+|  Use Global Mip Bias | Toggle   | On, Off         | Enables the global mipmap bias that Unity automatically imposes at runtime. Unity sets this bias during certain dynamic resolution scaling algorithms to improve detail reconstruction. |
```
The global bias gets applied on a per node basis. On by default. This bias results in the calculations of DRS. Ill make sure to explain this a bit better.

kecho replied:
ok redid this sentence:
Bias mode applies an offset bias to regular texture sampling. If "Use Global Mip Bias" is enabled, this bias gets added to the Global Mip Bias

kecho commented:
Current status of this PR:
* Ben approives of design.
* Waiting on @alindmanUnity 's design review
* Must sync with URP team and provide something for testing

kecho commented:
I have added 2 tests, the spheres on the right side :
![image](https://user-images.githubusercontent.com/789671/149008077-1c1dd14e-d3e6-4dc1-a2f7-20a4d68d9264.png)

The top one is a test for texture array sampler modes, gradient, standard, lod and bias.
The bottom one is for simple 2d texture, same modes.
Will now run this through the CI and get screenshots updated.

alindmanUnity commented on com.unity.shadergraph/Documentation~/Sample-Texture-2D-Array-Node.md:
```diff
@@ -16,6 +16,10 @@ NOTE: This [Node](Node.md) can only be used in the **Fragment** shader stage.
 | Index   | Input | Float       | None  | Index of array to sample |
 | UV      | Input | Vector 2    |   UV  | UV coordinates |
 | Sampler | Input | Sampler State | Default sampler state | Sampler for the texture |
+| LOD | Input | Float | LOD | Available when mip sampling mode is LOD. The actual mip to sample from. |
+| Bias | Input | Float | Bias | Available when mip sampling mode is Bias. A bias on top of the current global mip bias to sample from. |
```
this applies to the rest of the docs as well, but there is any docs in the main unity manual that explain what mip levels are that we can cross link to?

bencloward replied:
Sarah is working on this right now.

sarah-welton replied:
Can confirm, have something in progress that I will share shortly :)

kecho replied:
Tests are green. I am ready to merge this.

alindmanUnity reviewed (changes requested):
i'm on board UX wise but i have a question about the docs i'd like to address first
